In [1]:
import ast              # modulo para interpretación del json en caracteres
import json             # modulo json para interpretacion
import pandas as pd     # modulo panda para manejo de datos
import nltk             # Modulo para procesamiento de PLN


Mirando los archivos base, se decide que la mejor opcion para trabajar es por medio del modulo de pandas

# df_user_items

In [2]:
# creamos una lista que contenga los datos extraidos del archivo 'australian_user_items.json'
reviews = []
#abrimos el archivo json
with open('Dataset/australian_users_items.json', 'r', encoding='utf-8') as f:
    #creamos un loop para insertar los datos en una lista
    for line in f.readlines():
        reviews.append(ast.literal_eval(line))
#creamos un DataFrame con los datos
df_user_items = pd.DataFrame(reviews)

In [3]:
#mostramos la info del DataFrame
df_user_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88310 entries, 0 to 88309
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      88310 non-null  object
 1   items_count  88310 non-null  int64 
 2   steam_id     88310 non-null  object
 3   user_url     88310 non-null  object
 4   items        88310 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.4+ MB


In [4]:
#verificamos el contenido de la columna anidada 'items'
print(df_user_items['items'][0])

[{'item_id': '10', 'item_name': 'Counter-Strike', 'playtime_forever': 6, 'playtime_2weeks': 0}, {'item_id': '20', 'item_name': 'Team Fortress Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '30', 'item_name': 'Day of Defeat', 'playtime_forever': 7, 'playtime_2weeks': 0}, {'item_id': '40', 'item_name': 'Deathmatch Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '50', 'item_name': 'Half-Life: Opposing Force', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '60', 'item_name': 'Ricochet', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '70', 'item_name': 'Half-Life', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '130', 'item_name': 'Half-Life: Blue Shift', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '300', 'item_name': 'Day of Defeat: Source', 'playtime_forever': 4733, 'playtime_2weeks': 0}, {'item_id': '240', 'item_name': 'Counter-Strike: Source', 'playtime_forever': 1853, 'playtime_2weeks': 0}, {'item_i

Info:

    {'item_id': '10',

    'item_name': 'Counter-Strike',

    'playtime_forever': 6,

    'playtime_2weeks': 0}

In [5]:
#desanidamos utilizando lambda keys para recorrer los valores y generar las nuevas columnas
df_user_items['playtime_forever'] = df_user_items['items'].apply(lambda x: x[0].get('playtime_forever') if len(x) > 0 else 0)
df_user_items['playtime_2weeks'] = df_user_items['items'].apply(lambda x: x[0].get('playtime_2weeks') if len(x) > 0 else 0)
df_user_items['item_id'] = df_user_items['items'].apply(lambda x: x[0].get('item_id') if len(x) > 0 else 0)
df_user_items['playtime_forever'] = df_user_items['playtime_forever'].astype(int)
df_user_items['playtime_2weeks'] = df_user_items['playtime_2weeks'].astype(int)

In [6]:
#verificamos el resultado despues del desanidado
df_user_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88310 entries, 0 to 88309
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   user_id           88310 non-null  object
 1   items_count       88310 non-null  int64 
 2   steam_id          88310 non-null  object
 3   user_url          88310 non-null  object
 4   items             88310 non-null  object
 5   playtime_forever  88310 non-null  int32 
 6   playtime_2weeks   88310 non-null  int32 
 7   item_id           88310 non-null  object
dtypes: int32(2), int64(1), object(5)
memory usage: 4.7+ MB


In [7]:
#removemos las columnas que no voy a usar
df_user_items.drop(['user_url'], axis=1, inplace=True)
df_user_items.drop(['items'], axis=1, inplace=True)
df_user_items.drop(['items_count'], axis=1, inplace=True)

In [8]:
#mostramos la tabla final
df_user_items

user_id           steam_id  playtime_forever  \
0            76561197970982479  76561197970982479                 6   
1                      js41637  76561198035864385                 0   
2                    evcentric  76561198007712555               923   
3                   Riot-Punch  76561197963445855                 0   
4                        doctr  76561198002099482              1131   
...                        ...                ...               ...   
88305        76561198323066619  76561198323066619                 0   
88306        76561198326700687  76561198326700687                 0   
88307  XxLaughingJackClown77xX  76561198328759259                 0   
88308        76561198329548331  76561198329548331               677   
88309        edward_tremethick  76561198331598578                 0   

       playtime_2weeks item_id  
0                    0      10  
1                    0      10  
2                    0    1200  
3                    0      10  
4                    0     300  
...                ...     ...  
88305                0  413850  
88306                0   11020  
88307                0       0  
88308              677  304930  
88309                0       0  

[88310 rows x 5 columns]

# df_user_reviews

In [9]:
# realizamos el mismo procedimiento usado para procesar el json 'australian_users_items.json'
reviews = []
with open('Dataset/australian_user_reviews.json', 'r', encoding='utf-8') as f:
    for line in f.readlines():
        reviews.append(ast.literal_eval(line))
df_user_r = pd.DataFrame(reviews)

In [10]:
#verificamos el resultado despues del desanidado
df_user_r.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25799 entries, 0 to 25798
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   user_id   25799 non-null  object
 1   user_url  25799 non-null  object
 2   reviews   25799 non-null  object
dtypes: object(3)
memory usage: 604.8+ KB


In [11]:
#verificamos el contenido de la columna anidada 'items'
print(df_user_r['reviews'][0])

[{'funny': '', 'posted': 'Posted November 5, 2011.', 'last_edited': '', 'item_id': '1250', 'helpful': 'No ratings yet', 'recommend': True, 'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'}, {'funny': '', 'posted': 'Posted July 15, 2011.', 'last_edited': '', 'item_id': '22200', 'helpful': 'No ratings yet', 'recommend': True, 'review': "It's unique and worth a playthrough."}, {'funny': '', 'posted': 'Posted April 21, 2011.', 'last_edited': '', 'item_id': '43110', 'helpful': 'No ratings yet', 'recommend': True, 'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]


Info:

    {'funny': '',

    'posted': 'Posted November 5, 2011.',

    'last_edited': '',

    'item_id': '1250',
  
    'helpful': 'No ratings yet',

    'recommend': True,

    'review': 'Simple yet with......',}

     

In [12]:
#desanidamos la columna 'reviews' 
dfreviewsOpen = df_user_r.explode('reviews') 
df_user_reviews = pd.concat([dfreviewsOpen.drop(['reviews'], axis=1), dfreviewsOpen['reviews'].apply(pd.Series)], axis=1)

In [13]:
#borro  las columnas que no  necesito
df_user_reviews = df_user_reviews.drop(df_user_reviews.columns[-1], axis=1)
df_user_reviews.drop(['user_url', 'funny', 'last_edited'], axis=1, inplace=True)
#se crea la columna year
df_user_reviews['posted year'] = df_user_reviews['posted'].str.extract(r'(\d{4})')
#Se elimina la columna posted
df_user_reviews.drop('posted' , axis = 1, inplace = True)

In [14]:
#uso la libreria 'nltk' con 'vader_lexicon' y 'sentiment' para transformar los 'reviews' en 'score_reviews'
import nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer

#inicializo SentimentAnalyzer
sia = SentimentIntensityAnalyzer()

#funcion para asignar valores de acuerdo a la escala
def get_sentiment_score(text):
    if pd.isnull(text) or text == '':
        return 1  #return netral si la casilla esta vacia o contiene NaN
    elif isinstance(text, str):
        sentiment = sia.polarity_scores(text)
        compound_score = sentiment['compound']
        if compound_score >= -0.05:
            return 0  #buena opinion
        elif compound_score <= -0.05:
            return 2  #mala opinion
        else:
            return 1
    else:
        return 1  #return neutral para valores que no son string


#convierte la columna 'review' a str
df_user_reviews['review'] = df_user_reviews['review'].astype(str)

#aplica la funcion 'get_sentiment_score' a la columna 'review'
df_user_reviews['sentiment_score'] = df_user_reviews['review'].apply(get_sentiment_score)

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\carep\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [15]:
#miro la tabla resultante
df_user_reviews

user_id item_id  \
0      76561197970982479    1250   
0      76561197970982479   22200   
0      76561197970982479   43110   
1                js41637  251610   
1                js41637  227300   
...                  ...     ...   
25797  76561198312638244      70   
25797  76561198312638244  362890   
25798        LydiaMorley  273110   
25798        LydiaMorley     730   
25798        LydiaMorley     440   

                                               helpful recommend  \
0                                       No ratings yet      True   
0                                       No ratings yet      True   
0                                       No ratings yet      True   
1      15 of 20 people (75%) found this review helpful      True   
1         0 of 1 people (0%) found this review helpful      True   
...                                                ...       ...   
25797                                   No ratings yet      True   
25797                                   No ratings yet      True   
25798    1 of 2 people (50%) found this review helpful      True   
25798                                   No ratings yet      True   
25798                                   No ratings yet      True   

                                                  review posted year  \
0      Simple yet with great replayability. In my opi...        2011   
0                   It's unique and worth a playthrough.        2011   
0      Great atmosphere. The gunplay can be a bit chu...        2011   
1      I know what you think when you see this title ...        2014   
1      For a simple (it's actually not all that simpl...        2013   
...                                                  ...         ...   
25797  a must have classic from steam definitely wort...         NaN   
25797  this game is a perfect remake of the original ...         NaN   
25798  had so much fun plaing this and collecting res...         NaN   
25798                                                 :D         NaN   
25798                                     so much fun :D         NaN   

       sentiment_score  
0                    0  
0                    0  
0                    0  
1                    0  
1                    0  
...                ...  
25797                0  
25797                0  
25798                0  
25798                0  
25798                0  

[59333 rows x 7 columns]

In [16]:

#verifico los valores para 'sentiment_score' dentro del DataFrame
df_user_reviews.loc[df_user_reviews['sentiment_score'] == 2]

user_id item_id  \
3                    doctr  207610   
8        76561198089393905   72850   
14                DJKamBer    1250   
21     starkillershadow553  238090   
21     starkillershadow553  242130   
...                    ...     ...   
25772        iwishihadaids  391460   
25782    76561198270958927  381210   
25788           vinquility  413150   
25789    76561198289386531  304930   
25790          GodLoveGuru  242760   

                                              helpful recommend  \
3                                      No ratings yet      True   
8      3 of 3 people (100%) found this review helpful      True   
14                                     No ratings yet      True   
21      1 of 2 people (50%) found this review helpful      True   
21                                     No ratings yet      True   
...                                               ...       ...   
25772                                  No ratings yet     False   
25782    0 of 2 people (0%) found this review helpful      True   
25788                                  No ratings yet      True   
25789                                  No ratings yet      True   
25790  1 of 1 people (100%) found this review helpful      True   

                                                  review posted year  \
3      The ending to this game is.... ♥♥♥♥♥♥♥.... Jus...        2012   
8      Killed the Emperor, nobody cared and got away ...        2015   
14     Compared to Left 4 Dead 2, this game REALLY gi...        2013   
21     One of the best sequals i have every seen, reb...        2015   
21     If you're looking for a no frills, Arcade/Simu...         NaN   
...                                                  ...         ...   
25772                                                Bad         NaN   
25782  Fix your ♥♥♥♥ing game, de rank me for no reaso...         NaN   
25788  It's pretty fun, very very similiar to harvest...         NaN   
25789  I really like Unturned but the thing that i wa...         NaN   
25790  Good Game Play, encountered a problem with cli...         NaN   

       sentiment_score  
3                    2  
8                    2  
14                   2  
21                   2  
21                   2  
...                ...  
25772                2  
25782                2  
25788                2  
25789                2  
25790                2  

[9612 rows x 7 columns]

In [17]:
#grabo la tabla df_user_reviews como csv 
#df_user_reviews.to_csv('df_user_reviews.csv', index=False)

# df_steam_games

In [18]:
# realizamos el mismo procedimiento usado para procesar el json 'australian_users_items.json'
rows = []
with open ('Dataset/output_steam_games.json', 'r', encoding='utf-8') as f:
    for line in f.readlines():
        rows.append(json.loads(line))
df_s_games = pd.DataFrame(rows)

In [19]:
# elimino las columnas que no  necesito
df_s_games = df_s_games.drop(['publisher', 'app_name', 'url', 'tags', 'reviews_url', 'discount_price', 'specs', 'price', 'early_access', 'metascore', 'developer', 'items_count'], axis=1)


In [20]:
#verificamos el contenido de la columna anidada 'items'
print(df_s_games['items'][0])

[{'item_id': '10', 'item_name': 'Counter-Strike', 'playtime_forever': 6, 'playtime_2weeks': 0}, {'item_id': '20', 'item_name': 'Team Fortress Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '30', 'item_name': 'Day of Defeat', 'playtime_forever': 7, 'playtime_2weeks': 0}, {'item_id': '40', 'item_name': 'Deathmatch Classic', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '50', 'item_name': 'Half-Life: Opposing Force', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '60', 'item_name': 'Ricochet', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '70', 'item_name': 'Half-Life', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '130', 'item_name': 'Half-Life: Blue Shift', 'playtime_forever': 0, 'playtime_2weeks': 0}, {'item_id': '300', 'item_name': 'Day of Defeat: Source', 'playtime_forever': 4733, 'playtime_2weeks': 0}, {'item_id': '240', 'item_name': 'Counter-Strike: Source', 'playtime_forever': 1853, 'playtime_2weeks': 0}, {'item_i

Info , check Json keys

    {'item_id': '10', 
    
    'item_name': 'Counter-Strike', 
    
    'playtime_forever': 6, 
    
    'playtime_2weeks': 0}


In [21]:
#desanidamos utilizando lambda keys para recorrer los valores y generar las nuevas columnas
df_s_games['item_id'] = df_s_games['items'].apply(lambda x: x[0].get('item_id') if isinstance(x, list) and len(x) > 0 else 0)
df_s_games['item_name'] = df_s_games['items'].apply(lambda x: x[0].get('item_name') if isinstance(x, list) and len(x) > 0 else 0)
df_s_games['playtime_forever'] = df_s_games['items'].apply(lambda x: x[0].get('playtime_forever') if isinstance(x, list) and len(x) > 0 else 0)
df_s_games['playtime_2weeks'] = df_s_games['items'].apply(lambda x: x[0].get('playtime_2weeks') if isinstance(x, list) and len(x) > 0 else 0)
df_s_games['playtime_forever'] = df_s_games['playtime_forever'].astype(int)
df_s_games['playtime_2weeks'] = df_s_games['playtime_2weeks'].astype(int)

In [22]:
#remuevo  la columna que no voy a usar
df_s_games.drop(['items'], axis=1, inplace=True)

In [23]:
#se extrae el 'año' unicamente de la columna 'release date'
df_s_games['release_date'] = df_s_games['release_date'].str.extract(r'(\d{4})')


In [24]:
# Creamos dummies para genres
df_s_games['genres'] = df_s_games['genres'].fillna('[]')  # Rellenar los valores faltantes con una lista vacía
df_s_games['genres'] = df_s_games['genres'].apply(lambda x: ', '.join(x))  # Convertir la lista de géneros a una cadena separada por comas

# Crear variables ficticias para los géneros
dummy_genres = df_s_games['genres'].str.get_dummies(', ')

# Concatenar las variables ficticias con el DataFrame original
df_steam_games = pd.concat([df_s_games, dummy_genres], axis=1)

In [25]:
# borro  las columnas que no  necesito
df_steam_games = df_steam_games.drop(df_steam_games.columns[-1], axis=1)
df_steam_games = df_steam_games.drop(df_steam_games.columns[-1], axis=1)
df_steam_games.drop('genres' , axis = 1, inplace = True)
# normalizamos el nombre de algunas columnas
df_steam_games.rename(columns={'Animation &amp; Modeling': 'Animation and Modeling'}, inplace=True)
df_steam_games.rename(columns={'Design &amp; Illustration': 'Design and Illustration'}, inplace=True)
#elimino la columna 'item_id' pues la informacion que contiene no es la que me interesa
df_steam_games.drop(columns={'item_id'}, inplace=True)
#cambio el nombre de la columna 'id' que si tiene la informacion relevanto por 'item_id
df_steam_games.rename(columns={'id': 'item_id'}, inplace=True)

In [26]:
#miro la tabla final
df_steam_games

title release_date item_id            user_id  \
0                            NaN          NaN     NaN  76561197970982479   
1                            NaN          NaN     NaN            js41637   
2                            NaN          NaN     NaN          evcentric   
3                            NaN          NaN     NaN         Riot-Punch   
4                            NaN          NaN     NaN              doctr   
...                          ...          ...     ...                ...   
120440            Colony On Mars         2018  773640                NaN   
120441  LOGistICAL: South Africa         2018  733530                NaN   
120442             Russian Roads         2018  610660                NaN   
120443       EXIT 2 - Directions         2017  658870                NaN   
120444                       NaN          NaN  681550                NaN   

                 steam_id                      item_name  playtime_forever  \
0       76561197970982479                 Counter-Strike                 6   
1       76561198035864385                 Counter-Strike                 0   
2       76561198007712555  Red Orchestra: Ostfront 41-45               923   
3       76561197963445855                 Counter-Strike                 0   
4       76561198002099482          Day of Defeat: Source              1131   
...                   ...                            ...               ...   
120440                NaN                              0                 0   
120441                NaN                              0                 0   
120442                NaN                              0                 0   
120443                NaN                              0                 0   
120444                NaN                              0                 0   

        playtime_2weeks  Accounting  Action  ...  Photo Editing  RPG  Racing  \
0                     0           0       0  ...              0    0       0   
1                     0           0       0  ...              0    0       0   
2                     0           0       0  ...              0    0       0   
3                     0           0       0  ...              0    0       0   
4                     0           0       0  ...              0    0       0   
...                 ...         ...     ...  ...            ...  ...     ...   
120440                0           0       0  ...              0    0       0   
120441                0           0       0  ...              0    0       0   
120442                0           0       0  ...              0    0       1   
120443                0           0       0  ...              0    0       0   
120444                0           0       0  ...              0    0       0   

        Simulation  Software Training  Sports  Strategy  Utilities  \
0                0                  0       0         0          0   
1                0                  0       0         0          0   
2                0                  0       0         0          0   
3                0                  0       0         0          0   
4                0                  0       0         0          0   
...            ...                ...     ...       ...        ...   
120440           1                  0       0         1          0   
120441           0                  0       0         1          0   
120442           1                  0       0         0          0   
120443           0                  0       0         0          0   
120444           0                  0       0         0          0   

        Video Production  Web Publishing  
0                      0               0  
1                      0               0  
2                      0               0  
3                      0               0  
4                      0               0  
...                  ...             ...  
120440                 0               0  
120441                 0       

# df_merged
Uno  las tablas 'df_steam_games' y 'df_user_items' usando la columna key 'item_id'

In [27]:
#hago el merge por la columna 'item_id'
df_merged = df_steam_games.merge(df_user_items, on='item_id')

In [28]:
#borro las columnas sin informacion y las que no voy a usar
df_merged.drop(columns={'user_id_x'}, inplace=True)
df_merged.drop(columns={'item_name'}, inplace=True)
df_merged.drop(columns={'playtime_forever_x'}, inplace=True)
df_merged.drop(columns={'playtime_2weeks_x'}, inplace=True)
df_merged.drop(columns={'steam_id_y'}, inplace=True)
#normalizo el nombre de otras columnas
df_merged.rename(columns={'user_id_y': 'user_id'}, inplace=True)
df_merged.rename(columns={'playtime_forever_y': 'playtime_forever'}, inplace=True)
df_merged.rename(columns={'playtime_2weeks_y': 'playtime_2weeks'}, inplace=True)
df_merged.rename(columns={'steam_id_x': 'steam_id'}, inplace=True)

In [29]:
#miro  la tabla resultante
df_merged

title release_date item_id  \
0                                            Half-Life         1998      70   
1                                            Half-Life         1998      70   
2                                            Half-Life         1998      70   
3                                            Half-Life         1998      70   
4                                            Half-Life         1998      70   
...                                                ...          ...     ...   
63217  Unreal Tournament 2004: Editor's Choice Edition         2004   13230   
63218  Unreal Tournament 2004: Editor's Choice Edition         2004   13230   
63219  Unreal Tournament 2004: Editor's Choice Edition         2004   13230   
63220  Unreal Tournament 2004: Editor's Choice Edition         2004   13230   
63221  Unreal Tournament 2004: Editor's Choice Edition         2004   13230   

      steam_id  Accounting  Action  Adventure  Animation and Modeling  \
0          NaN           0       1          0                       0   
1          NaN           0       1          0                       0   
2          NaN           0       1          0                       0   
3          NaN           0       1          0                       0   
4          NaN           0       1          0                       0   
...        ...         ...     ...        ...                     ...   
63217      NaN           0       1          0                       0   
63218      NaN           0       1          0                       0   
63219      NaN           0       1          0                       0   
63220      NaN           0       1          0                       0   
63221      NaN           0       1          0                       0   

       Audio Production  Casual  ...  Simulation  Software Training  Sports  \
0                     0       0  ...           0                  0       0   
1                     0       0  ...           0                  0       0   
2                     0       0  ...           0                  0       0   
3                     0       0  ...           0                  0       0   
4                     0       0  ...           0                  0       0   
...                 ...     ...  ...         ...                ...     ...   
63217                 0       0  ...           0                  0       0   
63218                 0       0  ...           0                  0       0   
63219                 0       0  ...           0                  0       0   
63220                 0       0  ...           0                  0       0   
63221                 0       0  ...           0                  0       0   

       Strategy  Utilities  Video Production  Web Publishing  \
0             0          0                 0               0   
1             0          0                 0               0   
2             0          0                 0               0   
3             0          0                 0               0   
4             0          0                 0               0   
...         ...        ...               ...             ...   
63217         0          0                 0               0   
63218         0          0                 0               0   
63219         0          0                 0               0   
63220         0          0                 0               0   
63221         0          0                 0               0   

                 user_id  playtime_forever  playtime_2weeks  
0                kube134                 4                0  
1      76561198030567998                21                0  
2                itslonk                 8                0  
3      76561198056857968               451                0  
4             jonasdbomb                 0                0  
...                  ...               ...              ...  
63217  76561198071493773               368                0  
63218  76

# df_merged_final
Uno  las tablas 'df_merged' y 'df_user_reviews' usando la columna key 'user_id'

In [30]:
#al abrir el archivo df_user_reviews en PowerBI produce 16 lineas con errores.... 
#abrimos el  archivo en VISUAL STUDIO y borramos manualmente del archivo plano estas 16 lines buscando el siguiente texto 
#1 'BUT YEAH THIS GAME IS INCEREDIBLE WITH THE AMOUNT OF CHAARACHTERS '
#2 'Best driving game I have played'"Material de apoyo.md'
#3 'P.S It's in Beta so probs be fixed"'
#4 'The main problem is the online'
#5 'most of the gamemodes are fun to play'
#6 'A Must Get Game... It requires a game with the source engine (Half life 2'
#7 'Get counter strike source with it. It contains all the textures and models '
#8 'The animations need work'
#9 'Multiplayer is fun'
#10 'The realism is immersive and scary at times (You need to check up on ammo'
#11 'Its fun'
#12 'P.S It's in Beta so probs be fixed"'
#13 'The car revs wild everytime you get into it'
#14 'Must have game in your collection.. And if they would allow modding tools '
#15 '- None'
#16 '= It's a Zombie game. At least they aren't Nazis... /Senses incoming DLC./"'

In [31]:
#cargo desde el.csv el archivo 'df_user_reviews' que es el modificado y depurado
df_user_reviews = pd.read_csv(r"df_user_reviews.csv")

In [32]:
#merge 'df_merged' y 'df_user_reviews' por  la columna key 'user_id'
df_merged_final = df_merged.merge(df_user_reviews, on='user_id')

In [33]:
#borro las filas que tengan el campo 'user_id' vacio.
df_merged_final = df_merged_final.dropna(subset=['user_id'])
#normalizo el nombre de otras columnas
df_merged_final.rename(columns={'item_id_y': 'item_id'}, inplace=True)
df_merged_final.rename(columns={'item_id_x': 'id'}, inplace=True)


In [34]:
#miro la tabla resultante
df_merged_final

title release_date     id  \
0                                            Half-Life         1998     70   
1                                            Half-Life         1998     70   
2                                            Half-Life         1998     70   
3                                            Half-Life         1998     70   
4                                            Half-Life         1998     70   
...                                                ...          ...    ...   
53007                                    Day of Defeat         2003     30   
53008                                    Day of Defeat         2003     30   
53009                                    Day of Defeat         2003     30   
53010  Unreal Tournament 2004: Editor's Choice Edition         2004  13230   
53011  Unreal Tournament 2004: Editor's Choice Edition         2004  13230   

      steam_id  Accounting  Action  Adventure  Animation and Modeling  \
0          NaN           0       1          0                       0   
1          NaN           0       1          0                       0   
2          NaN           0       1          0                       0   
3          NaN           0       1          0                       0   
4          NaN           0       1          0                       0   
...        ...         ...     ...        ...                     ...   
53007      NaN           0       1          0                       0   
53008      NaN           0       1          0                       0   
53009      NaN           0       1          0                       0   
53010      NaN           0       1          0                       0   
53011      NaN           0       1          0                       0   

       Audio Production  Casual  ...  Web Publishing            user_id  \
0                     0       0  ...               0            kube134   
1                     0       0  ...               0  76561198030567998   
2                     0       0  ...               0  76561198030567998   
3                     0       0  ...               0  76561198030567998   
4                     0       0  ...               0  76561198030567998   
...                 ...     ...  ...             ...                ...   
53007                 0       0  ...               0  76561198085345522   
53008                 0       0  ...               0  76561198085345522   
53009                 0       0  ...               0     Shadowbolt4231   
53010                 0       0  ...               0  76561198045631910   
53011                 0       0  ...               0  76561198152281826   

       playtime_forever  playtime_2weeks  item_id  \
0                     4                0   251990   
1                    21                0   332800   
2                    21                0   319630   
3                    21                0     6020   
4                    21                0    55230   
...                 ...              ...      ...   
53007                 4                0     1250   
53008                 4                0   218230   
53009                15                0   293010   
53010               133                0   252490   
53011              2952               51    39140   

                                                 helpful  recommend  \
0         1 of 1 people (100%) found this review helpful       True   
1      194 of 282 people (69%) found this review helpful       True   
2          1 of 2 people (50%) found this review helpful       True   
3                                         No ratings yet       True   
4                                         No ratings yet       True   
...                                                  ...        ...   
53007                                     No ratings yet       True   
53008                                     No ratings yet       True   
53009      2 of 3 people (67%) found this review helpful  

In [35]:
#guardo esta tabla que es la fuente de informacion para todo el desarollo, como csv
df_merged_final.to_csv('df_merged_final.csv', index=False)

# Main.py
Creamos  las funciones de consulta

1.   PlayTimeGenre
2.   UserForGenre
3.   UsersRecommend
4.   UsersNotRecommend
5.   sentiment_analysis

In [29]:
df = pd.read_csv('df_merged_final.csv') 

In [30]:
from fastapi import FastAPI #se comenta ya que sino no corre los demas codigo y se habilita en el la carpeta de FastAPI - main.py
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics.pairwise import linear_kernel
from sklearn.feature_extraction.text import TfidfVectorizer


In [32]:
app=FastAPI(debug=True)

#abrimos el nuevo csv para realizar las funciones y hacer las pruebas
df = pd.read_csv('df_merged_final.csv') 

#mostramos un mensaje de bienvenida
@app.get('/')
def message():
  return 'Proyecto Individual Carlos Eduardo Peña'

#esta función devuelve el año de lanzamiento con más horas jugadas para un género determinado.
@app.get('/PlayTimeGenre/')
def PlayTimeGenre(genre: str) -> dict:
  #convierte el género a mayúscula inicial para que sea consistente con el nombre de las columnas del dataframe.
  genre = genre.capitalize()
  #filtra el dataframe para que solo contenga los juegos del género especificado
  #luego, calcula el total de horas jugadas por año y devuelve el año con el valor más alto.
  genre_df = df[df[genre] == 1]
  year_playtime_df = genre_df.groupby('year')['playtime_forever'].sum().reset_index()
  max_playtime_year = year_playtime_df.loc[year_playtime_df['playtime_forever'].idxmax(), 'year']
  #devuelve un diccionario con el género y el año de lanzamiento con más horas jugadas.
  return{'Genero': genre, 'Año de lanzamiento con mas horas jugadas para Genero:': int(max_playtime_year)}

#esta función devuelve el usuario con más horas jugadas para un género determinado.
@app.get('/UserForGenre/')
def UserForGenre(genre: str) -> dict:
  #convierte el género a mayúscula inicial para que sea consistente con el nombre de las columnas del dataframe.
  genre = genre.capitalize()
  #filtra el dataframe para que solo contenga los juegos del género especificado
  #luego, calcula el total de horas jugadas por usuario y devuelve el usuario con el valor más alto.
  genre_df = df[df[genre] == 1]
  max_playtime_user = genre_df.loc[genre_df['playtime_forever'].idxmax(), 'user_id']
  year_playtime_df = genre_df.groupby('year')['playtime_forever'].sum().reset_index()
  playtime_list = year_playtime_df.to_dict(orient='record')
  #devuelve un diccionario con el género, el usuario con más horas jugadas y una lista con las horas jugadas por año.
  result = {
      'Usuario con mas horas jugadas para Genero' + genre: max_playtime_user,
      'Horas jugadas': playtime_list}
  return result


#funciones `UsersRecommend()` y `UsersNotRecommend()
#estas funciones devuelven los tres juegos con la calificación más alta (2) para un año determinado
#la función `UsersRecommend()` solo considera los juegos que fueron recomendados por el usuario, 
#mientras que la función `UsersNotRecommend()` solo considera los juegos que no fueron recomendados por el usuario.
@app.get('/UsersRecommend/')
def UsersRecommend(year: int) -> dict:
  df_filtrado = df[(df['year']== year) & (df['recommend'] == True) & (df['sentiment_analysis'] == 2)]
  if df_filtrado.empty:
    return {'error': 'no encontrado'}
  df_ordenado = df_filtrado.sort_values(by='sentiment_analysis', ascending=False)
  top_3_resenas = df_ordenado.head(3)
  resultado = {
      'Puesto 1': top_3_resenas.iloc[0]['title'],
      'Puesto 2': top_3_resenas.iloc[1]['title'],
      'Puesto 3': top_3_resenas.iloc[2]['title'],
  }
  return resultado

@app.get('/UsersNotRecommend/')
def UsersNotRecommend(year: int)-> dict:
  df_filtrado = df[(df['year']== year) & (df['recommend'] == False) & (df['sentiment_analysis'] <=1)]
  if df_filtrado.empty:
    return {'error': 'no encontrado'}
  df_ordenado = df_filtrado.sort_values(by='sentiment_analysis', ascending=False)
  top_3_resenas = df_ordenado.head(3)
  resultado = {
      'Puesto 1': top_3_resenas.iloc[0]['title'],
      'Puesto 2': top_3_resenas.iloc[1]['title'],
      'Puesto 3': top_3_resenas.iloc[2]['title'],
  }
  return resultado

#esta función devuelve un diccionario con el número de juegos con cada calificación de sentimiento para un año determinado.
@app.get('/sentiment_analysis/')
def sentiment_analysis(year: int) -> dict:
  filtra_df = df[df['year']== year]
  sentiment = filtra_df['sentiment_analysis'].value_counts()
  result = {
      'Positivo': int(sentiment.get(2, 0)),
      'Neutral': int(sentiment.get(1, 0)),
      'Negativo': int(sentiment.get(0, 0))
  }
  return result


muestra = df.head(25000)
tfidf = TfidfVectorizer(stop_words='english')
muestra = muestra.fillna('')

tfidf_matri = tfidf.fit_transform(muestra['review'])
cosine_similarity = linear_kernel(tfidf_matri, tfidf_matri)


#esta función devuelve una lista con los tres juegos más similares a los juegos que el usuario ha jugado.
@app.get('/recomendacion_usuario/{id_juego}')
def recomendacion_juego(id_juego: int) -> dict:
    if id_juego not in muestra['id'].values:
        return {'mensaje': 'No existe el id del juego.'}
    titulo = muestra.loc[muestra['id']== id_juego, 'title'].iloc[0]
    idx = muestra[muestra['title'] == titulo].index[0]
    sim_cosine = list(enumerate(cosine_similarity[idx]))
    sim_scores = sorted(sim_cosine, key=lambda x: x[1], reverse=True)
    sim_ind = [i for i, _ in sim_scores[1:6]]
    sim_juegos = muestra['title'].iloc[sim_ind].values.tolist()
   
    return {'juegos recomendados': list(sim_juegos)}

probamos las funciones para realizar la consulta

In [33]:
df.columns #vemos las columnas del df

Index(['title', 'release_date', 'id', 'steam_id', 'Accounting', 'Action',
       'Adventure', 'Animation and Modeling', 'Audio Production', 'Casual',
       'Design and Illustration', 'Early Access', 'Education', 'Free to Play',
       'Indie', 'Massively Multiplayer', 'Photo Editing', 'RPG', 'Racing',
       'Simulation', 'Software Training', 'Sports', 'Strategy', 'Utilities',
       'Video Production', 'Web Publishing', 'user_id', 'playtime_forever',
       'playtime_2weeks', 'item_id', 'helpful', 'recommend', 'review',
       'posted year', 'sentiment_score'],
      dtype='object')

In [34]:
#miramos que años tienen informacion para ser usada en esta funcion
print(df['posted year'].unique())

[2014. 2015.   nan 2013. 2012. 2011. 2010.]


In [35]:
def sentiment_analysis(year: int) -> dict:
  filtra_df = df[df['posted year']== year] #filtramos el dataframe por año
  sentiment = filtra_df['sentiment_score'].value_counts() #contamos los valores de la columna sentiment_analysis
  result = {
      'Positivo': int(sentiment.get(2, 0)),
      'Neutral': int(sentiment.get(1, 0)),
      'Negativo': int(sentiment.get(0, 0))
  }
  return result

In [36]:
#llamamos a la funcion sentiment_analysis y guardamos el resultado en la variable resultado
resultado5 = sentiment_analysis(2012)
print(resultado5)
#Según el año de lanzamiento, se devuelve una lista con la cantidad de registros de
#reseñas de usuarios que se encuentren categorizados con un análisis de sentimiento.

{'Positivo': 142, 'Neutral': 0, 'Negativo': 990}


In [37]:
def UsersNotRecommend(year: int)-> dict:
  df_filtrado = df[(df['posted year']== year) & (df['recommend'] == False) & (df['sentiment_score'] <=1)]
  if df_filtrado.empty:
    return {'error': 'no encontrado'}
  df_ordenado = df_filtrado.sort_values(by='sentiment_score', ascending=False)
  if len(df_ordenado) < 3:
    return {'error': 'no hay suficientes reseñas'}
  top_3_resenas = df_ordenado.head(3)
  resultado = {
      'Puesto 1': top_3_resenas.iloc[0]['title'],
      'Puesto 2': top_3_resenas.iloc[1]['title'],
      'Puesto 3': top_3_resenas.iloc[2]['title'],
  }
  return resultado

In [38]:
#llamamos a la funcion UsersNotRecommend y guardamos el resultado en la variable resultado
resultado4 = UsersNotRecommend(2015)
print(resultado4)
#Devuelve el top 3 de juegos MENOS recomendados por usuarios para el año dado.

{'Puesto 1': "Garry's Mod", 'Puesto 2': 'Counter-Strike: Source', 'Puesto 3': 'Half-Life'}


In [39]:
def UsersRecommend(year: int) -> dict:
##filtramos el dataframe por año, recomendacion y sentimiento
  df_filtrado = df[(df['posted year']== year) & (df['recommend'] == True) & (df['sentiment_score'] == 2)]
  if df_filtrado.empty:
    return {'error': 'no encontrado'} #si el dataframe esta vacio devolvemos un error
  df_ordenado = df_filtrado.sort_values(by='sentiment_score', ascending=False) #ordenamos el dataframe por sentimiento
  if len(df_ordenado) < 3:
    return {'error': 'no hay suficientes reseñas'}
  top_3_resenas = df_ordenado.head(3) #seleccionamos las 3 primeras filas
  resultado = {
      'Puesto 1': top_3_resenas.iloc[0]['title'],
      'Puesto 2': top_3_resenas.iloc[1]['title'],
      'Puesto 3': top_3_resenas.iloc[2]['title'],
  }
  return resultado

In [40]:
#llamamos a la funcion UsersRecommend y guardamos el resultado en la variable resultado
resultado3 = UsersRecommend(2015)
print(resultado3)
#Devuelve el top 3 de juegos MÁS recomendados por usuarios para el año dado.

{'Puesto 1': 'Half-Life', 'Puesto 2': 'Counter-Strike: Source', 'Puesto 3': 'Counter-Strike: Source'}


In [41]:
#buscamos un genero con informacion para usar en esta funcion
df['Action'] 

0        1
1        1
2        1
3        1
4        1
        ..
53007    1
53008    1
53009    1
53010    1
53011    1
Name: Action, Length: 53012, dtype: int64

In [42]:
def UserForGenre(genre: str) -> dict:
    # convertimos el genero a mayusculas
    genre = genre.capitalize()
    # filtramos el dataframe por genero
    genre_df = df[df[genre] == 1]
    # seleccionamos el usuario con mas horas jugadas
    max_playtime_user = genre_df.loc[genre_df['playtime_forever'].idxmax(), 'user_id']
    # agrupamos el dataframe por año y sumamos las horas jugadas
    year_playtime_df = genre_df.groupby('posted year')['playtime_forever'].sum().reset_index()
    # convertimos el dataframe a diccionario con una entrada para cada fila
    playtime_list = year_playtime_df.to_dict(orient='records')
    result = {
        'Usuario con mas horas jugadas para Genero ' + genre: max_playtime_user,
        'Horas jugadas': playtime_list,
    }
    return result

In [43]:
#llamamos a la funcion UsersRecommend y guardamos el resultado en la variable resultado
resultado2 = UserForGenre('Action')
print(resultado2)
#Debe devolver el usuario que acumula más horas jugadas para el género
#dado y una lista de la acumulación de horas jugadas por año.

{'Usuario con mas horas jugadas para Genero Action': 'jiumohoo', 'Horas jugadas': [{'posted year': 2010.0, 'playtime_forever': 44637}, {'posted year': 2011.0, 'playtime_forever': 877869}, {'posted year': 2012.0, 'playtime_forever': 2363510}, {'posted year': 2013.0, 'playtime_forever': 8172851}, {'posted year': 2014.0, 'playtime_forever': 23092993}, {'posted year': 2015.0, 'playtime_forever': 20220112}]}


In [44]:
def PlayTimeGenre(genre: str) -> dict:
  genre = genre.capitalize() #convertimos el genero a mayusculas
  genre_df = df[df[genre] == 1] #filtramos el dataframe por genero
  year_playtime_df = genre_df.groupby('posted year')['playtime_forever'].sum().reset_index() #agrupamos el dataframe por año y sumamos las horas jugadas
  max_playtime_year = year_playtime_df.loc[year_playtime_df['playtime_forever'].idxmax(), 'posted year'] #seleccionamos el año con mas horas jugadas
  return{'Genero': genre, 'Año de lanzamiento con mas horas jugadas para ' + genre + ' es': int(max_playtime_year)} #devolvemos el genero y el año con mas horas jugadas

In [45]:
#llamamos a la funcion PlayTimeGenre y guardamos el resultado en la variable resultado
resultado = PlayTimeGenre('Action')
print(resultado)
#año con mas horas jugadas para dicho género

{'Genero': 'Action', 'Año de lanzamiento con mas horas jugadas para Action es': 2014}


In [46]:
#buscamos un genero con informacion para usar en esta funcion
df['id'] 

0           70
1           70
2           70
3           70
4           70
         ...  
53007       30
53008       30
53009       30
53010    13230
53011    13230
Name: id, Length: 53012, dtype: int64

In [47]:
muestra['id']

0         70
1         70
2         70
3         70
4         70
        ... 
24995    300
24996    300
24997    300
24998    300
24999    300
Name: id, Length: 25000, dtype: int64

In [48]:
def recomendacion_juego(id_juego: int) -> dict:
    if id_juego not in muestra['id'].values:
        return {'mensaje': 'No existe el id del juego.'}
    titulo = muestra.loc[muestra['id']== id_juego, 'title'].iloc[0]
    idx = muestra[muestra['title'] == titulo].index[0]
    sim_cosine = list(enumerate(cosine_similarity[idx]))
    sim_scores = sorted(sim_cosine, key=lambda x: x[1], reverse=True)
    sim_ind = [i for i, _ in sim_scores[1:6]]
    sim_juegos = muestra['title'].iloc[sim_ind].values.tolist()
   
    return {'juegos recomendados': list(sim_juegos)}

In [49]:
#llamamos a la funcion recomendacion_juego y guardamos el resultado en la variable resultado
resultado6 = recomendacion_juego(70)
print(resultado6)
#Devuelve 

{'juegos recomendados': ["Garry's Mod", "Garry's Mod", "Garry's Mod", "Garry's Mod", "Garry's Mod"]}
